packages/headless/src/features/insight-user-actions/insight-user-actions-preprocessing.ts

```diff
@@ -0,0 +1,349 @@
+    timestamp: rawAction.time,
+    eventData: {
+      type: actionData.event_type,
+      value: actionData.event_value,
+    },
+    cause: actionData.cause,
+    searchHub: actionData.origin_level_1,
+    document: {
+      title: actionData.title,
+      uriHash: actionData.uri_hash,
+      // eslint-disable-next-line @cspell/spellchecker
+      contentIdKey: actionData.c_contentidkey,
+      // eslint-disable-next-line @cspell/spellchecker
+      contentIdValue: actionData.c_contentidvalue,
+    },
+    query: actionData.query_expression,
+  };
+};
+
+/**
+ * Sorts by most recent and maps the raw user actions into user actions.
+ * We use reduce instead of map to drop any actions that could not be parsed.
+ * @param rawUserActions {rawUserAction[]} - The raw user actions to sort and map
+ * @returns {UserAction[]}
+ */
+export const mapAndSortActionsByMostRecent = (
+  rawUserActions: RawUserAction[]
+): UserAction[] => {
+  let parsingError = false;
+  const parsedActions = rawUserActions
+    .reduce((acc, rawAction) => {
+      try {
+        const userAction: UserAction = mapRawActionToUserAction(rawAction);
+        acc.push(userAction);
+      } catch (error) {
+        parsingError = true;
+      }
+      return acc;
+    }, [] as UserAction[])
+    .sort(
+      (firstAction, secondAction) =>
+        Number(secondAction.timestamp) - Number(firstAction.timestamp)
+    );
+
+  if (parsingError) {
+    console.warn(
+      'Some user actions could not be parsed. Please check the raw user actions data.'
+    );
+  }
+
+  return parsedActions;
+};
+
+/**
+ * Checks whether an action is within the same session as the previous action based on the specified session inactivity threshold.
+ * @param action {UserAction} - The current action
+ * @param previousTimestamp {string} - The timestamp of the previous action
+ * @returns {boolean}
+ */
+export const isActionWithinSessionThreshold = (
+  action: UserAction,
+  previousTimestamp: string
+): boolean => {
+  return (
+    Math.abs(Number(action.timestamp) - Number(previousTimestamp)) <
+    SESSION_INACTIVITY_THRESHOLD
+  );
+};
+
+/**
+ * Checks if the timestamp is within a specified range.
+ * @param timestamp {string} - The timestamp to check
+ * @param startTimestamp {number} - The start timestamp
+ * @param endTimestamp {number} - The end timestamp
+ * @returns {boolean}
+ */
+const isTimestampInRange = (
+  timestamp: string,
+  startTimestamp: string,
+  endTimestamp: string
+): boolean =>
+  Number(timestamp) >= Number(startTimestamp) &&
+  Number(timestamp) <= Number(endTimestamp);
+
+/**
+ * Inserts the case creation action at the right index in the current session actions array.
+ * @param currentSession {UserSession} - The current session
+ * @param caseCreationAction {UserAction} - The case creation action
+ * @returns {void}
+ */
+const insertCaseCreationActionInCurrentSession = (
+  currentSession: UserSession,
+  caseCreationAction: UserAction
```

Comment:
I had passed it there as that way I could declare it once and pass it to the method and lower return it as the returnTimline.session if it was not part of any session and since I need to pass the ticketCreationDate anyways if I dont so it doesn't really lower the # of params... Up to you, I can define it in the function and pass it later on outside the loop

Reply:
I dont know that we need to generalize here since "non-caseCreation" actions are pushed to the current session in order by most recent anyways in the prior step. The case creation action is the only one we really need to "insert" so I think this is fine.